proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+
+Priorities and messages are inherited by constraints stemming from another one by solving. For example, if we have ``IfUndischarged (Eq [a]) msg``, we would end up with a constraint ``Eq a`` where the message ``msg`` is kept.
+
+Finally, the solver has to be updated with the following new rules for inequality checks ::
+
+    [G] IfApartUnsafe a b no yes ==> yes
+    [W] IfApartUnsafe a b no yes
+          | if a and b are apart ==> no
+          | otherwise            ==> yes
+
+Indeed, we could see `IfApartUnsafe` as defined whenever compilation is successful as (again, using ``UndecidableSuperClasses``) ::
+
+    class yes => IfApartUnsafe a b no yes
+
+
+Effect and Interactions
+-----------------------
+In principle, there should be no effect for already-existing code. Right now people do not assume any specific ordering on the constraint solver, which means that we can change it as explained here.
+
+If implemented as-is, using ``(==)`` with the aforementioned type signature would impose a performance cost. Whereas before we just needed a ``Eq`` dictionary and a type application, now we need two of them and a complex proof of the big constraint. Luckily, this is easy to fix by using an ``INLINE`` pragma ::
+
+  {-# INLINE (==) #-}
+  (==) :: CustomErrors [ ... ] => a -> b -> Bool
+  (==) = eq
+
+where ``eq`` is the function with the simpler type signature. Preliminary research shows that this solves most of the problems, but no check has been done for interaction with other pragmas or optimizations.
+
+One problematic aspect of this way to encode custom errors is that the amount of type variables in a signature tends to grow (for ``(==)`` we have already duplicated it!). This might or might not be a problem, depending on who you ask.
+
+Finally, it seems that the approach we use for solving interacts in weird ways with ambiguity checking. In our prototype built over branch 8.3, with some complex usages of ``CustomErrors`` some signatures are marked as ambiguous (and thus require ``AllowAmbiguousTypes`` which are not). This might require improvements to the ambiguity checker, maybe making it aware of ``IfUndischarged`` and ``IfApart``.
+
+
+Costs and Drawbacks
+-------------------
+At first glance, the development and maintenance costs should be small.
+
+In principle, this extension should make libraries easier to learn. One problem remains, though: how to show type signatures in an understandable way. For most people ::
+
+    (==) :: Eq a => a -> a -> Bool
+
+conveys much more meaning that the long signature. Whether we should show one or the other in the interactive environment is discussed as an unresolved questions.
+
+One drawback of this approach is that it requires library authors to use ``DataKinds`` whenever they want to use custom errors. As a side-effect, some of the type definitions in a module might be promoted to the kind level. Right now we do not have a way to control data type promotion, so this is unavoidable.
```

Comment:
I don't see the drawback here. There is no difference between types and kinds anymore, so the problem isn't promoting a type to a kind -- the problem is promoting data constructors to be types. But because GHC looks for a proper (unpromoted) type before looking for a promoted data constructor, enabling `DataKinds` should never cause good code to fail to compile. Furthermore, whether or not these promoted types are available in downstream modules depends on the downstream modules' use (or not) of `DataKinds`.